syntax = "proto3";
package todo;

import "google/protobuf/timestamp.proto";

service TodoService {
    rpc SaveTodo(SaveTodoRequest) returns (TodoResponse) {}
    rpc ListTodo(ListTodoRequest) returns (TodoResponse) {}
    rpc ListAllTodo(ListAllTodoRequest) returns (stream ListAllTodoResponse) {}
    rpc DeleteTodo(DeleteTodoRequest) returns (TodoResponse) {}
    rpc EditTodo(EditTodoRequest) returns (TodoResponse) {}
    rpc ToggleStatus(ToggleStatusRequest) returns (TodoResponse) {}
}

message SaveTodoRequest{
    Todo todo = 1;
}

message ListTodoRequest {
    int32 id = 1;
}

message ListAllTodoRequest {}

message ListAllTodoResponse {
    OperationErrors status = 1;
    Todo todo = 2;
}

message DeleteTodoRequest {
    int32 id = 1;
}

message EditTodoRequest {
    Todo todo = 1;
}

message ToggleStatusRequest {
    int32 id = 1;
}

message TodoResponse {
    OperationErrors status = 1;
    Todo todo = 2;
}

message Todo {
    int32 id = 1;
    string title = 2;
    string description = 3;
    int32 status = 4;
    google.protobuf.Timestamp created_at = 5;
    google.protobuf.Timestamp updated_at = 6;
}

message OperationErrors {
    Status code = 1;
    repeated string errors = 2;
}

enum Status {
    OK = 0;
    ERROR = 1;
    NOT_FOUND = 2;
    UPDATED = 4;
    DELETED = 5;
}